homeassistant/helpers/script.py

```diff
@@ -220,287 +236,349 @@ def referenced_entities(self):
         self._referenced_entities = referenced
         return referenced
 
-    def run(self, variables=None, context=None):
+    def run(self, variables=None, context=None, logger=None, message_base=None):
         """Run script."""
         asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
+            self.async_run(variables, context, logger, message_base), self.hass.loop
         ).result()
 
-    async def async_run(
-        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
+    class _ScriptRun:
+        def __init__(
+            self,
+            hass: HomeAssistant,
+            parent: "Script",
+            variables: Optional[Sequence] = None,
+            context: Optional[Context] = None,
+            logger: Optional[logging.Logger] = None,
+            message_base: Optional[str] = None,
+        ) -> None:
+            self.hass = hass
+            self._parent = parent
+            self._variables = variables
+            self._context = context
+            self._logger = logger or _LOGGER
+            self._message_base = message_base or "Error executing script"
+            self._actions = {
+                ACTION_DELAY: self._async_delay,
+                ACTION_WAIT_TEMPLATE: self._async_wait_template,
+                ACTION_CHECK_CONDITION: self._async_check_condition,
+                ACTION_FIRE_EVENT: self._async_fire_event,
+                ACTION_CALL_SERVICE: self._async_call_service,
+                ACTION_DEVICE_AUTOMATION: self._async_device_automation,
+                ACTION_ACTIVATE_SCENE: self._async_activate_scene,
+            }
+            self.task: Optional[asyncio.Task] = None
+            self.last_action = None
+            self._cur = -1
+            self._async_listener: List[CALLBACK_TYPE] = []
+
+        async def async_run(self) -> None:
```

Comment:
We should greatly simplify this Instead of keeping track of `self._cur` etc, we should just `await asyncio.sleep(DELAY)`.

Reply:
Stopping a script should just cancel that task and we should catch `CancelledError`

Reply:
I had seriously considered doing just that, but that would cause even more breakage.

As it stands now, when a script is run via a service call (or the action part of an automation is run) it is done as a "blocking" call. However, the block only lasts until a delay or wait_template (that waits) is executed. As soon as the script gets to that point the service call "returns", and if that service call was part of another script (or automation action sequence), the caller continues to the next step asynchronously to the script that hit the delay/wait_template.

Using `asyncio.sleep` would mean the call (i.e., `await`) would continue to block, backing up the "call chain." This, I think, would be a _MAJOR_ behavioral change.

Rather than doing that at this point I'd suggest maybe considering allow scripts to be run completely asynchronously (as opposed to the current behavior which depends on whether or not the script contains a delay or wait_template.) Maybe that could be done as part of a future PR.

Reply:
Ok, we'll leave it for a future PR. I think that for users it would make more sense that automations wait till a script with a delay is fully done before continuing.

Reply:
> I think that for users it would make more sense that automations wait till a script with a delay is fully done before continuing.

I agree, although I've actually seen both sides. Some people assume calling a script (from an automation, another script, or wherever) would be asynchronous, and others assume it would be synchronous. Almost nobody assumes it works the way it does, until it's explained, and then they're like, huh?

I've also seen workarounds for both scenarios. I.e., to make calling a script fully asynchronous, the typical technique (knowing how they currently work) is to add a `delay: 0` as the first step of the script. And to make calling a script fully synchronous, an `input_boolean` is used -- it's set to `on` before calling the script, then a `wait_template` is used after the call to the script to wait until it's done, then inside the script, as the last step, the `input_boolean` is set to `off`, completing the `wait_template`. Neither workaround is pretty.

Actually I've been thinking about this some more and I think I have an even better idea (building on your suggestion.) But I haven't had the chance to really think it through, so I'll table it for now. But, at a minimum, to hopefully avoid, or at least minimize, future breakage, I'd like to make these changes now:

1. Move the `mode` parameter from `Script.__init__` to `Script.async_run` (which is where it's needed anyway), and change its name to `run_mode`.
1. Change the new `automation` parameter from `parallel_action` to `run_mode`. (I didn't like `parallel_action` anyway. 😃 )
1. Change the `allow` run mode option to `asynchronous` (in preparation for a possible `synchronous` mode for use in the `script` component.)

Reply:
I like the idea of a run mode, but we would still need a default run mode at automation and script integration level.

So changes that need to be made:

 1. introduce script helper run modes `blocking`, `background` and `background_after_delay` (legacy)
 2. add a new `default_run_mode` to scripts and automation actions. Default for both should initially be `background_after_delay` to mimic existing behavior.
 3. Add a new `run_mode` parameter when executing scripts
 4. If an automation is triggered while its action is still executing from a previous trigger, warn the user and don't trigger.
 5. Deal with automations being triggerd while last action is still executing. Allow strategies `restart`, `skip_delay` (legacy), `ignore`, `parallel`
 6. Clean up: remove all sync methods from the script helper and move `call_from_config` into the tests file.

I suggest that we close this PR and try to break it up in smaller pieces, each with tests.

Reply:
Ah, so a light has come on. 💡 There are really _two_ new parameters for `Script.async_run` (which your last comment finally made me see):

1. What to do if it's already running, _and_
1. when to return from the call.

I think I've been mushing them together in my mind.

The first could be called `if_running`, and it could have these choices: `"error"`, `"ignore"` (or `"skip"`), `"restart"` & `"parallel"` (or `"allow"`). The meanings are the same as we've been discussing.

The second could be called `blocking`, and it could have these choices: `True`, `False` & `"legacy"`.

`True`, of course, would cause `Script.async_run` not to return until the entire sequence is complete. In this case it would just run in the calling task's context. `False` would cause it to create a new task to run the sequence in and then return immediately.  `legacy` would run in the calling task's context until it hit a `delay` or a `wait_template` (that actually waits), at which time it would set up the rest of the sequence to run asynchronously and return (as it used to do.)

The `automation` integration would have a new config parameter `if_running` (or `mode` or whatever name sounds best) and would accept any of the values. The default would be `restart` if `action` contains a `delay` or `wait_template`, or `parallel` if it doesn't. But it would always just use `blocking=False`, because each trigger event runs in its own task, so there's no need to block. And, actually, it would want the call to `Script.async_run` to return immediately anyway so that it could update it's `last_triggered` attribute ASAP. (And `Script.async_run` would return `True` or `False` to indicate if the call resulted in a new run, and hence, `last_triggered` should be updated.)

The `script` integration would have both parameters (`if_running` and `blocking`), and accept all the values. Rather than make these part of the `script` configuration I think it makes more sense to be a parameter to the (service) call that initiates it running. That way it's up to the calling context to decide what makes sense. That gives the user the most flexibility. (Although it probably wouldn't be too difficult to convince me that it _should_ be part of the `script` configuration instead. lol) Defaults would be `if_running: error` and `blocking: legacy`, to maintain backwards compatibility. However it might also make sense to issue a deprecation warning for `legacy`.

One thing that would need to be addressed (which I've thought of but haven't mentioned until now) is the 10 second service call limit. To make truly blocking scripts possible it would have to not impose the timeout when the service call is to a script. The reason this was not a problem before is due to the behavior we've been discussing -- i.e., blocking only until a delay or wait. But, of course, that will no longer always be the case.

So, yeah, I think this is starting to sound pretty cool, but it does make sense to close this PR and do it in smaller chunks.

BTW, your item 4 would probably not be good, because that would actually cause breakage as I described previously. I.e., even though the script helper was faulty when it comes to parallel execution, in many cases (and especially without delays or waits) it actually "works" and, as I said, I think many existing automations depend on this.

Reply:
For 4, we will initially keep same behavior but warn if not explicitly specified that legacy behavior is wanted.

If we make the parameter part of script configuration, it would be to provide defaults . Still allowing it to be overridden via service call.